### PR TITLE
Provide a better support for Quarkus property value

### LINKED ISSUE #69: Provide a better support for Quarkus property value
Completion for properties value works for boolean (true, false) and enums. But for other case (ex : Level class for log Quarkus properties), it doesn't work.

The idea of this issue is to augment the properties values with a JSON settings (ex: for Level class,values are DEBUG, SEVERE, etc)

The Quarkus LS can provide a JSON settings by default about the properties values and user could extend it to add another values (ex : it adds a new Level).

### PR DESCRIPTION
**Pull request by angelozerr**
Provide a better support for Quarkus property value

Fixes #69

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR gives the capability to manage completion, hover and validation LEVEL for all properties which have value bound with java util logging Level.

I need to write tests and clean my code but you can play with it. See the following demo:

ValueRuleDemo

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+			return matchName(propertyName);
+		} else if (types != null) {
+			return matchType(propertyType);
+		}
+		// TODO: manage dependency properties (properties coming from the model).
+		return false;
+	}
+
+	private boolean matchName(String propertyName) {
+		return names.contains(propertyName);
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java]: check names is not null

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	private boolean matchType(String propertyType) {
+		return types.contains(propertyType);
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java]: check types is not null

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/ValuesRule.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	public void setValues(List<EnumItem> values) {
+		this.values = values;
+	}
+
+	public List<EnumItem> getValues() {
+		cleanValues();
+		return values;
+	}
+
+	private void cleanValues() {
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/ValuesRule.java]: should be done only once, in setValues(), might also want to make values immutable after cleaning everything

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java [hunk 4]
```diff
@@ -191,4 +200,13 @@ public Hover doHover(PropertiesModel document, Position position, QuarkusProject
 			QuarkusProjectInfo projectInfo, QuarkusFormattingSettings formattingSettings) {
 		return codeActions.doCodeActions(context, range, document, projectInfo, formattingSettings);
 	}
+
+	/**
+	 * Returns the manager for values rules.
+	 * 
+	 * @return the manager for values rules.
+	 */
+	private ValuesRulesManager getValuesRulesManger() {
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java]: getValuesRulesManager

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java [hunk 5]
```diff
@@ -50,15 +51,21 @@
 	private final QuarkusDiagnostics diagnostics;
 	private final QuarkusFormatter formatter;
 	private final QuarkusCodeActions codeActions;
+	private final ValuesRulesManager valuesRulesManger;
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java]: valuesRulesManager

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCompletionTest.java [hunk 6]
```diff
@@ -207,4 +220,14 @@ public void completionForExistingPropertiesDifferentProfile() throws BadLocation
 				c("quarkus.application.name", "%prod.quarkus.application.name=", r(2, 0, 6)));
 
 	}
+
+	@Test
+	public void completionOnValueForLevelBasedOnRule() throws BadLocationException {
+		// quarkus.log.file.level has 'java.util.logging.Level' which jave not
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCompletionTest.java]: which has no

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java [hunk 7]
```diff
@@ -400,14 +400,23 @@ public void validateFloatError() throws BadLocationException {
...
 				"quarkus.ssl.native=    ${value-three}";
-
 		QuarkusValidationSettings settings = new QuarkusValidationSettings();
-
 		testDiagnosticsFor(value, getDefaultQuarkusProjectInfo(), settings);
 	}
 
+	@Test
+	public void validateValueForLevelBasedOnRule() throws BadLocationException {
+		// quarkus.log.file.level has 'java.util.logging.Level' which jave not
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java]: which has no

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesHoverTest.java [hunk 8]
```diff
@@ -165,4 +170,20 @@ public void hoverWithEnums() throws BadLocationException {
 		String hoverLabel = "**BLOCK**" + System.lineSeparator();
 		assertHoverMarkdown(value, hoverLabel, 35);
 	}
+
+	@Test
+	public void hoverOnValueForLevelBasedOnRule() throws BadLocationException {
+		// quarkus.log.file.level has 'java.util.logging.Level' which jave not
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesHoverTest.java]: which has no

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java (hunk 2 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java]: It is done before.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/PropertyMatcher.java]: It is done before.

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesHoverTest.java (hunk 8 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesHoverTest.java]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java (hunk 7 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCompletionTest.java (hunk 6 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesCompletionTest.java]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java (hunk 5 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java (hunk 4 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java [hunk 9]
```diff
@@ -131,17 +131,58 @@ public void setEnums(List<EnumItem> enums) {
...
+	/**
+	 * Returns the enumeration item from the given property value and given
+	 * enumerations and null otherwise.
+	 * 
+	 * @param propertyValue the property value.
+	 * @param enums         enumerations.
+	 * @return the enumeration item from the given property value and given
+	 *         enumerations and null otherwise.
+	 */
+	public static EnumItem getEnumItem(String propertyValue, List<EnumItem> enums) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java [hunk 10]
```diff
@@ -131,17 +131,58 @@ public void setEnums(List<EnumItem> enums) {
...
+	 * @return the enumeration item from the given property value and given
+	 *         enumerations and null otherwise.
+	 */
+	public static EnumItem getEnumItem(String propertyValue, List<EnumItem> enums) {
 		if (enums == null) {
 			return null;
 		}
 		for (EnumItem enumItem : enums) {
-			if (enumItem.getName().equals(propertyValue)) {
+			if (enumItem.getName() != null && enumItem.getName().equals(propertyValue)) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java [hunk 11]
```diff
@@ -131,17 +131,58 @@ public void setEnums(List<EnumItem> enums) {
...
+	 * Returns the enumeration item from the given property value and given
+	 * enumerations and null otherwise.
+	 * 
+	 * @param propertyValue the property value.
+	 * @param enums         enumerations.
+	 * @return the enumeration item from the given property value and given
+	 *         enumerations and null otherwise.
+	 */
+	public static EnumItem getEnumItem(String propertyValue, List<EnumItem> enums) {
 		if (enums == null) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java]: @fbricon updated

**Comment by fbricon** [comment]
Doesn't work for [`quarkus.log.category.{*}.level`](https://lordofthejars.github.io/quarkus-cheat-sheet/#_logging)

**Comment by angelozerr** [comment]
> Doesn't work for quarkus.log.category.{*}.level

Yes I know, I don't know why but those property are String and not Level type.

See for instance https://quarkus.io/guides/logging#quarkus-core-log-config_quarkus.log.category.-categories-.min-level

If we can consider that it's the same than level, we should add this rule in the JSON by their name.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/ValuesRule.java (hunk 3 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/model/values/ValuesRule.java]: I'm using GSON to populate this values and it uses field, in other words setValues is never called.

**Comment by fbricon** [comment]
> If we can consider that it's the same than level, we should add this rule in the JSON by their name.

if it's supported for maps then yes

**Comment by angelozerr** [comment]
> if it's supported for maps then yes

@fbricon it should be supported and test has been modified with this feature.

**Comment by angelozerr** [comment]
@fbricon is there something you want I do ?

File: quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+					"name": "read-uncommitted"
+				},
+				{
+					"name": "read-committed"
+				},
+				{
+					"name": "serializable"
+				},
+				{
+					"name": "serializable"
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json]: duplicate value

File: quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+			},
+			"valuesRef": [
+				"#logginglevel"
+			]
+		},
+		{
+			"matcher": {
+				"names": [
+					"quarkus.log.category.{*}.level",
+					"quarkus.log.category.{*}.min-level"
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json]: when applying the templates, we insert:
```
quarkus.log.category.key.level=key
"quarkus.log.category.key.min-level=key
```
I expect the injected values to be the actual default values, or at the very list something from the list of acceptable values

File: quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json (hunk 12 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json (hunk 13 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/model/values/quarkus-values-rules.json]: Indeed there were a bug with index. Now it should work.

**Pull request merged**